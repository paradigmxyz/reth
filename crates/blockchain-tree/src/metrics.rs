use metrics::Histogram;
use reth_metrics::{
    metrics::{Counter, Gauge},
    Metrics,
};
use std::time::{Duration, Instant};

/// Metrics for the blockchain tree block buffer
#[derive(Metrics)]
#[metrics(scope = "blockchain_tree.block_buffer")]
pub struct BlockBufferMetrics {
    /// Total blocks in the block buffer
    pub blocks: Gauge,
}

#[derive(Debug)]
pub(crate) struct MakeCanonicalDurationsRecorder {
    start: Instant,
    pub(crate) actions: Vec<(MakeCanonicalAction, Duration)>,
    latest: Option<Duration>,
    current_metrics: MakeCanonicalMetrics,
}

impl Default for MakeCanonicalDurationsRecorder {
    fn default() -> Self {
        Self {
            start: Instant::now(),
            actions: Vec::new(),
            latest: None,
            current_metrics: MakeCanonicalMetrics::default(),
        }
    }
}

impl MakeCanonicalDurationsRecorder {
    /// Records the duration since last record, saves it for future logging and instantly reports as
    /// a metric with `action` label.
    pub(crate) fn record_relative(&mut self, action: MakeCanonicalAction) {
        let elapsed = self.start.elapsed();
        let duration = elapsed - self.latest.unwrap_or_default();

        self.actions.push((action, duration));
        self.current_metrics.record(action, duration);
        self.latest = Some(elapsed);
    }
}

/// Metrics for the entire blockchain tree
#[derive(Metrics)]
#[metrics(scope = "blockchain_tree")]
pub struct TreeMetrics {
    /// Total number of sidechains (not including the canonical chain)
    pub sidechains: Gauge,
    /// The highest block number in the canonical chain
    pub canonical_chain_height: Gauge,
    /// The number of reorgs
    pub reorgs: Counter,
    /// The latest reorg depth
    pub latest_reorg_depth: Gauge,
    /// Longest sidechain height
    pub longest_sidechain_height: Gauge,
    /// The number of times cached trie updates were used for insert.
    pub trie_updates_insert_cached: Counter,
    /// The number of times trie updates were recomputed for insert.
    pub trie_updates_insert_recomputed: Counter,
}

/// Represents actions for making a canonical chain.
#[derive(Debug, Copy, Clone)]
pub(crate) enum MakeCanonicalAction {
    /// Cloning old blocks for canonicalization.
    CloneOldBlocks,
    /// Finding the canonical header.
    FindCanonicalHeader,
    /// Splitting the chain for canonicalization.
    SplitChain,
    /// Splitting chain forks for canonicalization.
    SplitChainForks,
    /// Merging all chains for canonicalization.
    MergeAllChains,
    /// Updating the canonical index during canonicalization.
    UpdateCanonicalIndex,
    /// Retrieving (cached or recomputed) state trie updates
    RetrieveStateTrieUpdates,
    /// Committing the canonical chain to the database.
    CommitCanonicalChainToDatabase,
    /// Reverting the canonical chain from the database.
    RevertCanonicalChainFromDatabase,
    /// Inserting an old canonical chain.
    InsertOldCanonicalChain,
    /// Clearing trie updates of other children chains after fork choice update.
    ClearTrieUpdatesForOtherChildren,
}

/// Canonicalization metrics
#[derive(Metrics)]
#[metrics(scope = "blockchain_tree.make_canonical")]
struct MakeCanonicalMetrics {
    /// Duration of the clone old blocks action.
    clone_old_blocks: Histogram,
    /// Duration of the find canonical header action.
    find_canonical_header: Histogram,
    /// Duration of the split chain action.
    split_chain: Histogram,
    /// Duration of the split chain forks action.
    split_chain_forks: Histogram,
    /// Duration of the merge all chains action.
    merge_all_chains: Histogram,
    /// Duration of the update canonical index action.
    update_canonical_index: Histogram,
    /// Duration of the retrieve state trie updates action.
    retrieve_state_trie_updates: Histogram,
    /// Duration of the commit canonical chain to database action.
    commit_canonical_chain_to_database: Histogram,
    /// Duration of the revert canonical chain from database action.
    revert_canonical_chain_from_database: Histogram,
    /// Duration of the insert old canonical chain action.
    insert_old_canonical_chain: Histogram,
    /// Duration of the clear trie updates of other children chains after fork choice update
    /// action.
    clear_trie_updates_for_other_children: Histogram,
}

impl MakeCanonicalMetrics {
    /// Records the duration for the given action.
    pub(crate) fn record(&self, action: MakeCanonicalAction, duration: Duration) {
        match action {
            MakeCanonicalAction::CloneOldBlocks => self.clone_old_blocks.record(duration),
            MakeCanonicalAction::FindCanonicalHeader => self.find_canonical_header.record(duration),
            MakeCanonicalAction::SplitChain => self.split_chain.record(duration),
            MakeCanonicalAction::SplitChainForks => self.split_chain_forks.record(duration),
            MakeCanonicalAction::MergeAllChains => self.merge_all_chains.record(duration),
            MakeCanonicalAction::UpdateCanonicalIndex => {
                self.update_canonical_index.record(duration)
            }
            MakeCanonicalAction::RetrieveStateTrieUpdates => {
                self.retrieve_state_trie_updates.record(duration)
            }
            MakeCanonicalAction::CommitCanonicalChainToDatabase => {
                self.commit_canonical_chain_to_database.record(duration)
            }
            MakeCanonicalAction::RevertCanonicalChainFromDatabase => {
                self.revert_canonical_chain_from_database.record(duration)
            }
            MakeCanonicalAction::InsertOldCanonicalChain => {
                self.insert_old_canonical_chain.record(duration)
            }
            MakeCanonicalAction::ClearTrieUpdatesForOtherChildren => {
                self.clear_trie_updates_for_other_children.record(duration)
            }
        }
    }
}
