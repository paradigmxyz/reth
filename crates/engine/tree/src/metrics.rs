use reth_metrics::{
    metrics::{Gauge, Histogram},
    Metrics,
};

/// Metrics for the `BasicBlockDownloader`.
#[derive(Metrics)]
#[metrics(scope = "consensus.engine.beacon")]
pub(crate) struct BlockDownloaderMetrics {
    /// How many blocks are currently being downloaded.
    pub(crate) active_block_downloads: Gauge,
}

/// Metrics for the `PersistenceService`
#[derive(Metrics)]
#[metrics(scope = "consensus.engine.persistence")]
pub(crate) struct PersistenceMetrics {
    /// How long it took for blocks to be removed
    pub(crate) remove_blocks_above_duration_seconds: Histogram,
    /// How long it took for blocks to be saved
    pub(crate) save_blocks_duration_seconds: Histogram,
    /// How long it took for blocks to be pruned
    pub(crate) prune_before_duration_seconds: Histogram,
}
