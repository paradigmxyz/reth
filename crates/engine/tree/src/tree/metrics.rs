use reth_metrics::{
    metrics::{Counter, Gauge, Histogram},
    Metrics,
};

/// Metrics for the `EngineApi`.
#[derive(Metrics)]
#[metrics(scope = "consensus.engine.beacon")]
pub(crate) struct EngineApiMetrics {
    /// How many executed blocks are currently stored.
    pub(crate) executed_blocks: Gauge,
    /// The number of times the pipeline was run.
    pub(crate) pipeline_runs: Counter,
    /// The total count of forkchoice updated messages received.
    pub(crate) forkchoice_updated_messages: Counter,
    /// The total count of new payload messages received.
    pub(crate) new_payload_messages: Counter,
    /// Histogram of persistence operation durations (in seconds)
    pub(crate) persistence_duration: Histogram,
    // TODO add latency metrics
}
