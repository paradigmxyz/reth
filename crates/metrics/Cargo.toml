[package]
name = "reth-metrics"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
homepage.workspace = true
repository.workspace = true
description = "reth metrics utilities"

[lints]
workspace = true

[dependencies]
# reth
reth-metrics-derive.workspace = true

# metrics
metrics.workspace = true

# async
tokio = { workspace = true, features = ["full"], optional = true }
futures = { workspace = true, optional = true }
tokio-util = { workspace = true, optional = true }

[features]
common = ["tokio", "futures", "tokio-util"]
