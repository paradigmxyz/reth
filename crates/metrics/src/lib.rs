//! Collection of metrics utilities.
//!
//! ## Feature Flags
//!
//! - `common`: Common metrics utilities, such as wrappers around tokio senders and receivers. Pulls
//!   in `tokio`.

#![doc(
    html_logo_url = "https://raw.githubusercontent.com/paradigmxyz/reth/main/assets/reth-docs.png",
    html_favicon_url = "https://avatars0.githubusercontent.com/u/97369466?s=256",
    issue_tracker_base_url = "https://github.com/paradigmxyz/reth/issues/"
)]
#![cfg_attr(not(test), warn(unused_crate_dependencies))]
#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]

/// Metrics derive macro.
pub use reth_metrics_derive::Metrics;

/// Implementation of common metric utilities.
#[cfg(feature = "common")]
pub mod common;

/// Re-export core metrics crate.
pub use metrics;
