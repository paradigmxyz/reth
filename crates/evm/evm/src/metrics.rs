//! Executor metrics.
//!
//! Block processing related to syncing should take care to update the metrics by using either
//! [`ExecutorMetrics::execute_metered`] or [`ExecutorMetrics::metered_one`].
use crate::{Database, OnStateHook};
use alloy_consensus::BlockHeader;
use alloy_evm::{
    block::{BlockExecutor, StateChangeSource},
    Evm,
};
use core::borrow::BorrowMut;
use metrics::{Counter, Gauge, Histogram};
use reth_execution_errors::BlockExecutionError;
use reth_execution_types::BlockExecutionOutput;
use reth_metrics::Metrics;
use reth_primitives_traits::{Block, BlockBody, RecoveredBlock};
use revm::{
    database::{states::bundle_state::BundleRetention, State},
    state::EvmState,
};
use std::time::Instant;

/// Wrapper struct that combines metrics and state hook
struct MeteredStateHook {
    metrics: ExecutorMetrics,
    inner_hook: Box<dyn OnStateHook>,
}

impl OnStateHook for MeteredStateHook {
    fn on_state(&mut self, source: StateChangeSource, state: &EvmState) {
        // Update the metrics for the number of accounts, storage slots and bytecodes loaded
        let accounts = state.keys().len();
        let storage_slots = state.values().map(|account| account.storage.len()).sum::<usize>();
        let bytecodes = state
            .values()
            .filter(|account| !account.info.is_empty_code_hash())
            .collect::<Vec<_>>()
            .len();

        self.metrics.accounts_loaded_histogram.record(accounts as f64);
        self.metrics.storage_slots_loaded_histogram.record(storage_slots as f64);
        self.metrics.bytecodes_loaded_histogram.record(bytecodes as f64);

        // Call the original state hook
        self.inner_hook.on_state(source, state);
    }
}

/// Executor metrics.
// TODO(onbjerg): add sload/sstore
#[derive(Metrics, Clone)]
#[metrics(scope = "sync.execution")]
pub struct ExecutorMetrics {
    /// The total amount of gas processed.
    pub gas_processed_total: Counter,
    /// The instantaneous amount of gas processed per second.
    pub gas_per_second: Gauge,
    /// The Histogram for amount of gas used.
    pub gas_used_histogram: Histogram,

    /// The Histogram for amount of time taken to execute blocks.
    pub execution_histogram: Histogram,
    /// The total amount of time it took to execute the latest block.
    pub execution_duration: Gauge,

    /// The Histogram for number of accounts loaded when executing the latest block.
    pub accounts_loaded_histogram: Histogram,
    /// The Histogram for number of storage slots loaded when executing the latest block.
    pub storage_slots_loaded_histogram: Histogram,
    /// The Histogram for number of bytecodes loaded when executing the latest block.
    pub bytecodes_loaded_histogram: Histogram,

    /// The Histogram for number of accounts updated when executing the latest block.
    pub accounts_updated_histogram: Histogram,
    /// The Histogram for number of storage slots updated when executing the latest block.
    pub storage_slots_updated_histogram: Histogram,
    /// The Histogram for number of bytecodes updated when executing the latest block.
    pub bytecodes_updated_histogram: Histogram,
}

impl ExecutorMetrics {
    fn metered<F, R, B>(&self, block: &RecoveredBlock<B>, f: F) -> R
    where
        F: FnOnce() -> R,
        B: reth_primitives_traits::Block,
    {
        // Execute the block and record the elapsed time.
        let execute_start = Instant::now();
        let output = f();
        let execution_duration = execute_start.elapsed().as_secs_f64();

        // Update gas metrics.
        self.gas_processed_total.increment(block.header().gas_used());
        self.gas_per_second.set(block.header().gas_used() as f64 / execution_duration);
        self.gas_used_histogram.record(block.header().gas_used() as f64);
        self.execution_histogram.record(execution_duration);
        self.execution_duration.set(execution_duration);

        output
    }

    /// Execute the given block using the provided [`BlockExecutor`] and update metrics for the
    /// execution.
    ///
    /// Compared to [`Self::metered_one`], this method additionally updates metrics for the number
    /// of accounts, storage slots and bytecodes loaded and updated.
    /// Execute the given block using the provided [`BlockExecutor`] and update metrics for the
    /// execution.
    pub fn execute_metered<E, DB>(
        &self,
        executor: E,
        input: &RecoveredBlock<impl Block<Body: BlockBody<Transaction = E::Transaction>>>,
        state_hook: Box<dyn OnStateHook>,
    ) -> Result<BlockExecutionOutput<E::Receipt>, BlockExecutionError>
    where
        DB: Database,
        E: BlockExecutor<Evm: Evm<DB: BorrowMut<State<DB>>>>,
    {
        // clone here is cheap, all the metrics are Option<Arc<_>>. additionally
        // they are globally registered so that the data recorded in the hook will
        // be accessible.
        let wrapper = MeteredStateHook { metrics: self.clone(), inner_hook: state_hook };

        let mut executor = executor.with_state_hook(Some(Box::new(wrapper)));

        // Use metered to execute and track timing/gas metrics
        let (mut db, result) = self.metered(input, || {
            executor.apply_pre_execution_changes()?;
            for tx in input.transactions_recovered() {
                executor.execute_transaction(tx)?;
            }
            executor.finish().map(|(evm, result)| (evm.into_db(), result))
        })?;

        // merge transactions into bundle state
        db.borrow_mut().merge_transitions(BundleRetention::Reverts);
        let output = BlockExecutionOutput { result, state: db.borrow_mut().take_bundle() };

        // Update the metrics for the number of accounts, storage slots and bytecodes updated
        let accounts = output.state.state.len();
        let storage_slots =
            output.state.state.values().map(|account| account.storage.len()).sum::<usize>();
        let bytecodes = output.state.contracts.len();

        self.accounts_updated_histogram.record(accounts as f64);
        self.storage_slots_updated_histogram.record(storage_slots as f64);
        self.bytecodes_updated_histogram.record(bytecodes as f64);

        Ok(output)
    }

    /// Execute the given block and update metrics for the execution.
    pub fn metered_one<F, R, B>(&self, input: &RecoveredBlock<B>, f: F) -> R
    where
        F: FnOnce(&RecoveredBlock<B>) -> R,
        B: reth_primitives_traits::Block,
    {
        self.metered(input, || f(input))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloy_eips::eip7685::Requests;
    use alloy_evm::EthEvm;
    use alloy_primitives::{B256, U256};
    use metrics_util::debugging::{DebugValue, DebuggingRecorder, Snapshotter};
    use reth_ethereum_primitives::{Receipt, TransactionSigned};
    use reth_execution_types::BlockExecutionResult;
    use revm::{
        database::State,
        database_interface::EmptyDB,
        inspector::NoOpInspector,
        state::{Account, AccountInfo, AccountStatus, EvmStorage, EvmStorageSlot},
        Context, MainBuilder, MainContext,
    };
    use std::sync::mpsc;

    /// A mock executor that simulates state changes
    struct MockExecutor {
        state: EvmState,
        hook: Option<Box<dyn OnStateHook>>,
        evm: EthEvm<State<EmptyDB>, NoOpInspector>,
    }

    impl MockExecutor {
        fn new(state: EvmState) -> Self {
            let db = State::builder()
                .with_database(EmptyDB::default())
                .with_bundle_update()
                .without_state_clear()
                .build();
            let evm = EthEvm::new(
                Context::mainnet().with_db(db).build_mainnet_with_inspector(NoOpInspector {}),
                false,
            );
            Self { state, hook: None, evm }
        }
    }

    impl BlockExecutor for MockExecutor {
        type Transaction = TransactionSigned;
        type Receipt = Receipt;
        type Evm = EthEvm<State<EmptyDB>, NoOpInspector>;

        fn apply_pre_execution_changes(&mut self) -> Result<(), BlockExecutionError> {
            Ok(())
        }

        fn execute_transaction_with_result_closure(
            &mut self,
            _tx: impl alloy_evm::block::ExecutableTx<Self>,
            _f: impl FnOnce(&revm::context::result::ExecutionResult<<Self::Evm as Evm>::HaltReason>),
        ) -> Result<u64, BlockExecutionError> {
            Ok(0)
        }

        fn finish(
            self,
        ) -> Result<(Self::Evm, BlockExecutionResult<Self::Receipt>), BlockExecutionError> {
            let Self { evm, hook, .. } = self;

            // Call hook with our mock state
            if let Some(mut hook) = hook {
                hook.on_state(StateChangeSource::Transaction(0), &self.state);
            }

            Ok((
                evm,
                BlockExecutionResult {
                    receipts: vec![],
                    requests: Requests::default(),
                    gas_used: 0,
                },
            ))
        }

        fn set_state_hook(&mut self, hook: Option<Box<dyn OnStateHook>>) {
            self.hook = hook;
        }

        fn evm(&self) -> &Self::Evm {
            &self.evm
        }

        fn evm_mut(&mut self) -> &mut Self::Evm {
            &mut self.evm
        }
    }

    struct ChannelStateHook {
        output: i32,
        sender: mpsc::Sender<i32>,
    }

    impl OnStateHook for ChannelStateHook {
        fn on_state(&mut self, _source: StateChangeSource, _state: &EvmState) {
            let _ = self.sender.send(self.output);
        }
    }

    fn setup_test_recorder() -> Snapshotter {
        let recorder = DebuggingRecorder::new();
        let snapshotter = recorder.snapshotter();
        recorder.install().unwrap();
        snapshotter
    }

    #[test]
    fn test_executor_metrics_hook_metrics_recorded() {
        let snapshotter = setup_test_recorder();
        let metrics = ExecutorMetrics::default();
        let input = RecoveredBlock::<reth_ethereum_primitives::Block>::default();

        let (tx, _rx) = mpsc::channel();
        let expected_output = 42;
        let state_hook = Box::new(ChannelStateHook { sender: tx, output: expected_output });

        let state = {
            let mut state = EvmState::default();
            let storage =
                EvmStorage::from_iter([(U256::from(1), EvmStorageSlot::new(U256::from(2)))]);
            state.insert(
                Default::default(),
                Account {
                    info: AccountInfo {
                        balance: U256::from(100),
                        nonce: 10,
                        code_hash: B256::random(),
                        code: Default::default(),
                    },
                    storage,
                    status: AccountStatus::Loaded,
                },
            );
            state
        };
        let executor = MockExecutor::new(state);
        let _result = metrics.execute_metered::<_, EmptyDB>(executor, &input, state_hook).unwrap();

        let snapshot = snapshotter.snapshot().into_vec();

        for metric in snapshot {
            let metric_name = metric.0.key().name();
            if metric_name == "sync.execution.accounts_loaded_histogram" ||
                metric_name == "sync.execution.storage_slots_loaded_histogram" ||
                metric_name == "sync.execution.bytecodes_loaded_histogram"
            {
                if let DebugValue::Histogram(vs) = metric.3 {
                    assert!(
                        vs.iter().any(|v| v.into_inner() > 0.0),
                        "metric {metric_name} not recorded"
                    );
                }
            }
        }
    }

    #[test]
    fn test_executor_metrics_hook_called() {
        let metrics = ExecutorMetrics::default();
        let input = RecoveredBlock::<reth_ethereum_primitives::Block>::default();

        let (tx, rx) = mpsc::channel();
        let expected_output = 42;
        let state_hook = Box::new(ChannelStateHook { sender: tx, output: expected_output });

        let state = EvmState::default();

        let executor = MockExecutor::new(state);
        let _result = metrics.execute_metered::<_, EmptyDB>(executor, &input, state_hook).unwrap();

        let actual_output = rx.try_recv().unwrap();
        assert_eq!(actual_output, expected_output);
    }
}
