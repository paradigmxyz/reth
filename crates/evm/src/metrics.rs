//! Executor metrics.
//!
//! Block processing related to syncing should take care to update the metrics by using either
//! [`ExecutorMetrics::execute_metered`] or [`ExecutorMetrics::metered_one`].
use crate::{execute::Executor, system_calls::OnStateHook};
use metrics::{Counter, Gauge, Histogram};
use reth_execution_types::{BlockExecutionInput, BlockExecutionOutput};
use reth_metrics::Metrics;
use reth_primitives::BlockWithSenders;
use revm_primitives::ResultAndState;
use std::time::Instant;

/// Wrapper struct that combines metrics and state hook
struct MeteredStateHook {
    metrics: ExecutorMetrics,
    inner_hook: Box<dyn OnStateHook>,
}

impl OnStateHook for MeteredStateHook {
    fn on_state(&mut self, result_and_state: &ResultAndState) {
        // Update the metrics for the number of accounts, storage slots and bytecodes loaded
        let accounts = result_and_state.state.keys().len();
        let storage_slots =
            result_and_state.state.values().map(|account| account.storage.len()).sum::<usize>();
        let bytecodes = result_and_state
            .state
            .values()
            .filter(|account| !account.info.is_empty_code_hash())
            .collect::<Vec<_>>()
            .len();

        self.metrics.accounts_loaded_histogram.record(accounts as f64);
        self.metrics.storage_slots_loaded_histogram.record(storage_slots as f64);
        self.metrics.bytecodes_loaded_histogram.record(bytecodes as f64);

        // Call the original state hook
        self.inner_hook.on_state(result_and_state);
    }
}

/// Executor metrics.
// TODO(onbjerg): add sload/sstore
#[derive(Metrics, Clone)]
#[metrics(scope = "sync.execution")]
pub struct ExecutorMetrics {
    /// The total amount of gas processed.
    pub gas_processed_total: Counter,
    /// The instantaneous amount of gas processed per second.
    pub gas_per_second: Gauge,

    /// The Histogram for amount of time taken to execute blocks.
    pub execution_histogram: Histogram,
    /// The total amount of time it took to execute the latest block.
    pub execution_duration: Gauge,

    /// The Histogram for number of accounts loaded when executing the latest block.
    pub accounts_loaded_histogram: Histogram,
    /// The Histogram for number of storage slots loaded when executing the latest block.
    pub storage_slots_loaded_histogram: Histogram,
    /// The Histogram for number of bytecodes loaded when executing the latest block.
    pub bytecodes_loaded_histogram: Histogram,

    /// The Histogram for number of accounts updated when executing the latest block.
    pub accounts_updated_histogram: Histogram,
    /// The Histogram for number of storage slots updated when executing the latest block.
    pub storage_slots_updated_histogram: Histogram,
    /// The Histogram for number of bytecodes updated when executing the latest block.
    pub bytecodes_updated_histogram: Histogram,
}

impl ExecutorMetrics {
    fn metered<F, R>(&self, block: &BlockWithSenders, f: F) -> R
    where
        F: FnOnce() -> R,
    {
        // Execute the block and record the elapsed time.
        let execute_start = Instant::now();
        let output = f();
        let execution_duration = execute_start.elapsed().as_secs_f64();

        // Update gas metrics.
        self.gas_processed_total.increment(block.gas_used);
        self.gas_per_second.set(block.gas_used as f64 / execution_duration);
        self.execution_histogram.record(execution_duration);
        self.execution_duration.set(execution_duration);

        output
    }

    /// Execute the given block using the provided [`Executor`] and update metrics for the
    /// execution.
    ///
    /// Compared to [`Self::metered_one`], this method additionally updates metrics for the number
    /// of accounts, storage slots and bytecodes loaded and updated.
    /// Execute the given block using the provided [`Executor`] and update metrics for the
    /// execution.
    pub fn execute_metered<'a, E, DB, O, Error>(
        &self,
        executor: E,
        input: BlockExecutionInput<'a, BlockWithSenders>,
        state_hook: Box<dyn OnStateHook>,
    ) -> Result<BlockExecutionOutput<O>, Error>
    where
        E: Executor<
            DB,
            Input<'a> = BlockExecutionInput<'a, BlockWithSenders>,
            Output = BlockExecutionOutput<O>,
            Error = Error,
        >,
    {
        // clone here is cheap, all the metrics are Option<Arc<_>>. additionally
        // they are gloally registered so that the data recorded in the hook will
        // be accessible.
        let wrapper = MeteredStateHook { metrics: self.clone(), inner_hook: state_hook };

        // Store reference to block for metered
        let block = input.block;

        // Use metered to execute and track timing/gas metrics
        let output = self.metered(block, || executor.execute_with_state_hook(input, wrapper))?;

        // Update the metrics for the number of accounts, storage slots and bytecodes updated
        let accounts = output.state.state.len();
        let storage_slots =
            output.state.state.values().map(|account| account.storage.len()).sum::<usize>();
        let bytecodes = output.state.contracts.len();

        self.accounts_updated_histogram.record(accounts as f64);
        self.storage_slots_updated_histogram.record(storage_slots as f64);
        self.bytecodes_updated_histogram.record(bytecodes as f64);

        Ok(output)
    }

    /// Execute the given block and update metrics for the execution.
    pub fn metered_one<F, R>(&self, input: BlockExecutionInput<'_, BlockWithSenders>, f: F) -> R
    where
        F: FnOnce(BlockExecutionInput<'_, BlockWithSenders>) -> R,
    {
        self.metered(input.block, || f(input))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloy_eips::eip7685::Requests;
    use metrics_util::debugging::{DebugValue, DebuggingRecorder, Snapshotter};
    use revm::db::BundleState;
    use revm_primitives::{
        Account, AccountInfo, AccountStatus, Bytes, EvmState, EvmStorage, EvmStorageSlot,
        ExecutionResult, Output, SuccessReason, B256, U256,
    };
    use std::sync::mpsc;

    /// A mock executor that simulates state changes
    struct MockExecutor {
        result_and_state: ResultAndState,
    }

    impl Executor<()> for MockExecutor {
        type Input<'a>
            = BlockExecutionInput<'a, BlockWithSenders>
        where
            Self: 'a;
        type Output = BlockExecutionOutput<()>;
        type Error = std::convert::Infallible;

        fn execute(self, _input: Self::Input<'_>) -> Result<Self::Output, Self::Error> {
            Ok(BlockExecutionOutput {
                state: BundleState::default(),
                receipts: vec![],
                requests: Requests::default(),
                gas_used: 0,
            })
        }
        fn execute_with_state_closure<F>(
            self,
            _input: Self::Input<'_>,
            _state: F,
        ) -> Result<Self::Output, Self::Error>
        where
            F: FnMut(&revm::State<()>),
        {
            Ok(BlockExecutionOutput {
                state: BundleState::default(),
                receipts: vec![],
                requests: Requests::default(),
                gas_used: 0,
            })
        }
        fn execute_with_state_hook<F>(
            self,
            _input: Self::Input<'_>,
            mut hook: F,
        ) -> Result<Self::Output, Self::Error>
        where
            F: OnStateHook + 'static,
        {
            // Call hook with our mock state
            hook.on_state(&self.result_and_state);

            Ok(BlockExecutionOutput {
                state: BundleState::default(),
                receipts: vec![],
                requests: Requests::default(),
                gas_used: 0,
            })
        }
    }

    struct ChannelStateHook {
        output: i32,
        sender: mpsc::Sender<i32>,
    }

    impl OnStateHook for ChannelStateHook {
        fn on_state(&mut self, _result_and_state: &ResultAndState) {
            let _ = self.sender.send(self.output);
        }
    }

    fn setup_test_recorder() -> Snapshotter {
        let recorder = DebuggingRecorder::new();
        let snapshotter = recorder.snapshotter();
        recorder.install().unwrap();
        snapshotter
    }

    #[test]
    fn test_executor_metrics_hook_metrics_recorded() {
        let snapshotter = setup_test_recorder();
        let metrics = ExecutorMetrics::default();

        let input = BlockExecutionInput {
            block: &BlockWithSenders::default(),
            total_difficulty: Default::default(),
        };

        let (tx, _rx) = mpsc::channel();
        let expected_output = 42;
        let state_hook = Box::new(ChannelStateHook { sender: tx, output: expected_output });

        let result_and_state = ResultAndState {
            result: ExecutionResult::Success {
                reason: SuccessReason::Stop,
                gas_used: 100,
                output: Output::Call(Bytes::default()),
                logs: vec![],
                gas_refunded: 0,
            },
            state: {
                let mut state = EvmState::default();
                let storage =
                    EvmStorage::from_iter([(U256::from(1), EvmStorageSlot::new(U256::from(2)))]);
                state.insert(
                    Default::default(),
                    Account {
                        info: AccountInfo {
                            balance: U256::from(100),
                            nonce: 10,
                            code_hash: B256::random(),
                            code: Default::default(),
                        },
                        storage,
                        status: AccountStatus::Loaded,
                    },
                );
                state
            },
        };
        let executor = MockExecutor { result_and_state };
        let _result = metrics.execute_metered(executor, input, state_hook).unwrap();

        let snapshot = snapshotter.snapshot().into_vec();

        for metric in snapshot {
            let metric_name = metric.0.key().name();
            if metric_name == "sync.execution.accounts_loaded_histogram" ||
                metric_name == "sync.execution.storage_slots_loaded_histogram" ||
                metric_name == "sync.execution.bytecodes_loaded_histogram"
            {
                if let DebugValue::Histogram(vs) = metric.3 {
                    assert!(
                        vs.iter().any(|v| v.into_inner() > 0.0),
                        "metric {metric_name} not recorded"
                    );
                }
            }
        }
    }

    #[test]
    fn test_executor_metrics_hook_called() {
        let metrics = ExecutorMetrics::default();

        let input = BlockExecutionInput {
            block: &BlockWithSenders::default(),
            total_difficulty: Default::default(),
        };

        let (tx, rx) = mpsc::channel();
        let expected_output = 42;
        let state_hook = Box::new(ChannelStateHook { sender: tx, output: expected_output });

        let result_and_state = ResultAndState {
            result: ExecutionResult::Revert { gas_used: 0, output: Default::default() },
            state: EvmState::default(),
        };
        let executor = MockExecutor { result_and_state };
        let _result = metrics.execute_metered(executor, input, state_hook).unwrap();

        let actual_output = rx.try_recv().unwrap();
        assert_eq!(actual_output, expected_output);
    }
}
