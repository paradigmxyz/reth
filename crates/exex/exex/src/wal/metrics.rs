use metrics::Gauge;
use reth_metrics::Metrics;

/// Metrics for the [WAL](`super::Wal`)
#[derive(Metrics)]
#[metrics(scope = "exex.wal")]
pub(super) struct Metrics {
    /// Size of all notifications in WAL in bytes
    pub size_bytes: Gauge,
    /// Number of notifications in WAL
    pub notifications_count: Gauge,
    /// Number of committed blocks in WAL
    pub committed_blocks_count: Gauge,
    /// Lowest committed block height in WAL
    pub lowest_committed_block_height: Gauge,
    /// Highest committed block height in WAL
    pub highest_committed_block_height: Gauge,
}
