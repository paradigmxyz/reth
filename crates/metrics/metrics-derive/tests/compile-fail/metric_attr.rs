extern crate metrics;
extern crate reth_metrics_derive;

use metrics::Gauge;
use reth_metrics_derive::Metrics;

fn main() {}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics {
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics2 {
    #[metric()]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics3 {
    #[metric(random = "value")]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics4 {
    #[metric(describe = 123)]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics5 {
    #[metric(rename = 123)]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics6 {
    #[metric(describe = "", describe = "")]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics7 {
    #[metric(rename = "_gauge", rename = "_gauge")]
    gauge: Gauge,
}

#[derive(Metrics)]
#[metrics(scope = "some_scope")]
struct CustomMetrics8 {
    #[metric(describe = "")]
    gauge: String,
}
