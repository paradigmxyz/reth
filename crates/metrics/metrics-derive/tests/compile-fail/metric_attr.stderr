error: Either doc comment or `describe = ..` must be set.
  --> tests/compile-fail/metric_attr.rs:12:5
   |
12 |     gauge: Gauge,
   |     ^^^^^^^^^^^^

error: Either doc comment or `describe = ..` must be set.
  --> tests/compile-fail/metric_attr.rs:18:5
   |
18 | /     #[metric()]
19 | |     gauge: Gauge,
   | |________________^

error: Unsupported attribute entry.
  --> tests/compile-fail/metric_attr.rs:25:14
   |
25 |     #[metric(random = "value")]
   |              ^^^^^^^^^^^^^^^^

error: Value **must** be a string literal.
  --> tests/compile-fail/metric_attr.rs:32:25
   |
32 |     #[metric(describe = 123)]
   |                         ^^^

error: Value **must** be a string literal.
  --> tests/compile-fail/metric_attr.rs:39:23
   |
39 |     #[metric(rename = 123)]
   |                       ^^^

error: Duplicate `describe` value provided.
  --> tests/compile-fail/metric_attr.rs:46:29
   |
46 |     #[metric(describe = "", describe = "")]
   |                             ^^^^^^^^^^^^^

error: Duplicate `rename` value provided.
  --> tests/compile-fail/metric_attr.rs:53:33
   |
53 |     #[metric(rename = "_gauge", rename = "_gauge")]
   |                                 ^^^^^^^^^^^^^^^^^

error: Unsupported metric type
  --> tests/compile-fail/metric_attr.rs:61:12
   |
61 |     gauge: String,
   |            ^^^^^^
